### PR TITLE
Kazoo: added basic kemi support

### PR DESCRIPTION
**Pull request by tao-communications**
#### Pre-Submission Checklist

- [*] Commit message has the format required by CONTRIBUTING guide
- [*] Commits are split per component (core, individual modules, libs, utils, ...)
- [*] Each component has a single commit (if not, squash them into one commit)
- [*] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [*] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [ ] PR should be backported to stable branches
- [*] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description

Added kemi support for
Parameter:
- event_callback
Functions:
- kazoo_subscribe(json_description)  
- kazoo_publish
Documentation:
- added section for event_callback parameter

### CODE REVIEW

**Comment by miconda** [comment]
Thanks, really appreciated being a module I am not using much, so I haven't approach it from KEMI perspective yet!

@lazedo - can you quickly check this PR and see if there is something you want to comment on?

Otherwise, I am fine to merge.

**Comment by lazedo** [comment]
@miconda i'll try to take a deeper look next week. at first glance,  some formatting issues and use of route_no which can break things already in place.

File: src/modules/kazoo/kazoo.c [hunk 1]
```diff
@@ -273,6 +279,12 @@ static int mod_init(void) {
    		return -1;
     }
 
+    if(faked_msg_init()<0)
+    {
+        LM_ERR("failed to init kazoo local sip msg\n");
+        return -1;
+    }
+
```
Comment (lazedo) [inline src/modules/kazoo/kazoo.c:287]: why was this added ?

File: src/modules/kazoo/kazoo.c [hunk 2]
```diff
@@ -312,7 +324,30 @@ static int mod_init(void) {
...
+			  return -1;
+		  }
+		  if (event_rt.rlist[route_no]==0)
+		  {
+			  LM_WARN("event_route[kazoo:consumer-event] is empty\n");
+		  }
+		  kazoo_route_no=route_no;
+    }
+    
+
```
Comment (lazedo) [inline src/modules/kazoo/kazoo.c:350]: please use a different logic here. `kazoo:consumer-event` is not mandatory. you could set a variable like `kz_use_kemi` that is set if `kazoo_event_callback` is set and `kemi` is valid. then use that variable when needed

File: src/modules/kazoo/kazoo.c [hunk 3]
```diff
@@ -435,27 +487,46 @@ static int fire_init_event(int rank)
...
-		init_run_actions_ctx(&ctx);
-		run_top_route(event_rt.rlist[rt], fmsg, &ctx);
-		if(ctx.run_flags&DROP_R_F)
-		{
-			LM_ERR("exit due to 'drop' in event route\n");
-			return -1;
+	if (kazoo_route_no>=0) {
+		rt = route_get(&event_rt, "kazoo:mod-init");
+        if(rt>=0 && event_rt.rlist[rt]!=NULL) {
+			LM_DBG("executing event_route[kazoo:mod-init] (%d)\n", rt);
```
Comment (lazedo) [inline src/modules/kazoo/kazoo.c:499]: formatting looks odd

File: src/modules/kazoo/kz_amqp.c [hunk 4]
```diff
@@ -2362,30 +2378,48 @@ void kz_amqp_consumer_event(kz_amqp_consumer_delivery_ptr Evt)
...
+            msg = faked_msg_next();
+            if(sr_kemi_route(keng, msg, EVENT_ROUTE, &kazoo_event_callback, &evrtname)<0)                {
+                LM_ERR("error running event route kemi callback\n");
+            }
+            set_route_type(rtb);
+        } else {
+            LM_ERR("no event route or kemi callback found for execution\n");
+        }
     }
+
```
Comment (lazedo) [inline src/modules/kazoo/kz_amqp.c:2422]: split this in two different functions that are evoked depending on `kemi` being active

File: src/modules/kazoo/kazoo.c [hunk 5]
```diff
@@ -435,27 +487,46 @@ static int fire_init_event(int rank)
...
+			str evrtname = str_init("kazoo:mod-init");
+			rtb = get_route_type();
+			msg = faked_msg_next();
+			if(sr_kemi_route(keng, msg, EVENT_ROUTE, &kazoo_event_callback, &evrtname)<0) {
+				LM_ERR("error running event route kemi callback\n");
+		  }
+		  set_route_type(rtb);
+		} 
+    else {
+			LM_ERR("no event route or kemi callback found for execution\n");
```
Comment (lazedo) [inline src/modules/kazoo/kazoo.c:528]: split this in two different function that are evoked by `fire_init_event` depending on `kemi` being active

**Comment by tao-communications** [comment]
I have addressed the suggested changes in the new commit:
    - removed route_lookup for kazoo:consumer-event as it may or may not exist
    - depending on using cfg or kemi:
     - fire_init_event() calls fire_init_event_cfg() or fire_init_event_kemi()
     - kz_amqp_consumer_event() calls kz_amqp_consumer_event_cfg() or
and tidied up formatting.

Tested on my test servers for: kazoo_subscribe,  kazoo_publish, kazoo:consumer-event, with the traditional cfg and Kemi (Lua).
    kz_amqp_consumer_event_kemi()

File: src/modules/kazoo/kazoo.c (hunk 1 above)
Comment (tao-communications) [reply src/modules/kazoo/kazoo.c]: This has been removed.

File: src/modules/kazoo/kazoo.c (hunk 2 above)
Comment (tao-communications) [reply src/modules/kazoo/kazoo.c]: Now removed route_lookup for kazoo:consumer-event as it may or may not exist.

File: src/modules/kazoo/kazoo.c (hunk 3 above)
Comment (tao-communications) [reply src/modules/kazoo/kazoo.c]: Now fixed formatting.

File: src/modules/kazoo/kz_amqp.c (hunk 4 above)
Comment (tao-communications) [reply src/modules/kazoo/kz_amqp.c]: Now depending on using cfg or kemi:
     - kz_amqp_consumer_event() calls kz_amqp_consumer_event_cfg() or kz_amqp_consumer_event_kemi()

File: src/modules/kazoo/kazoo.c (hunk 5 above)
Comment (tao-communications) [reply src/modules/kazoo/kazoo.c]: Now depending on using cfg or kemi:
     - fire_init_event() calls fire_init_event_cfg() or fire_init_event_kemi()

**Comment by lazedo** [comment]
@tao-communications great job, thank you. i do have a final request, when that's done, please squash your commits into a single commit, `kazoo: add basic kemi support`

i feel that `kazoo_route_no` doesn't express well the intent for the variable which is to use `kemi` instead of `cfg`.

can you please rename `kazoo_route_no`  to `kazoo_kemi_enabled` with default 0 and set it to 1 in  [here](https://github.com/kamailio/kamailio/pull/2324/files#diff-52d37ad2407d869cdcee8ffc29166e62R330)
and then use 
```
if ( kazoo_kemi_enabled )  {
   do kemi routine
} else {
   do cfg routine
}
```
where applicable, for example in `fire_init_event`
```
	if (kazoo_kemi_enabled) {
		return fire_init_event_kemi();
	}
	else {
		return fire_init_event_cfg();
	}
``

**Comment by tao-communications** [comment]
Thanks for reviewing! I made a new commit to address these.
Please let me know if there are further things that need doing.

**Comment by lazedo** [comment]
@tao-communications all good. please squash your commits so it can be merged

**Comment by miconda** [comment]
@lazedo: @tao-communications does not have commit access yet.

You can squash all commits via github merge button, just click on arrow down next to `Merge pull request` and choose `Squash and merge`. It is no need to have the squash done elsewhere.

I can do it if you prefer.

**Comment by lazedo** [comment]
@miconda i know that, i just asked to squash 6 commits into one. waiting on this and i will `Rebase and merge`. i want to preserve the original commit and not do a `merge`. thank you

**Comment by lazedo** [comment]
@miconda why did you dismissed my review ?

**Comment by miconda** [comment]
@lazedo - I dismissed the review because it was requesting an external squash, which is not necessary in my opinion. You can set the any commit message when you squash and merge via github.com web UI.

**Comment by tao-communications** [comment]
I have attempted a git squash and hopefully it's OK.

**Comment by lazedo** [comment]
@tao-communications looks good, just waiting for travis to complete. thanks.

**Comment by lazedo** [comment]
@tao-communications thank you again for your contribution.

**Comment by tao-communications** [comment]
Thank you very much @lazedo and @miconda for the review and support!
